// cql-execution/lib/elm/logical.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsFalse = exports.IsTrue = exports.Xor = exports.Not = exports.Or = exports.And = void 0;
const expression_1 = require("./expression");
const datatypes_1 = require("../datatypes/datatypes");
class And extends expression_1.Expression {
    constructor(json) {
        super(json);
    }
    exec(ctx) {
        return datatypes_1.ThreeValuedLogic.and(...this.execArgs(ctx));
    }
}
exports.And = And;
class Or extends expression_1.Expression {
    constructor(json) {
        super(json);
    }
    exec(ctx) {
        return datatypes_1.ThreeValuedLogic.or(...this.execArgs(ctx));
    }
}
exports.Or = Or;
class Not extends expression_1.Expression {
    constructor(json) {
        super(json);
    }
    exec(ctx) {
        return datatypes_1.ThreeValuedLogic.not(this.execArgs(ctx));
    }
}
exports.Not = Not;
class Xor extends expression_1.Expression {
    constructor(json) {
        super(json);
    }
    exec(ctx) {
        return datatypes_1.ThreeValuedLogic.xor(...this.execArgs(ctx));
    }
}
exports.Xor = Xor;
class IsTrue extends expression_1.Expression {
    constructor(json) {
        super(json);
    }
    exec(ctx) {
        return true === this.execArgs(ctx);
    }
}
exports.IsTrue = IsTrue;
class IsFalse extends expression_1.Expression {
    constructor(json) {
        super(json);
    }
    exec(ctx) {
        return false === this.execArgs(ctx);
    }
}
exports.IsFalse = IsFalse;
