// cql-translation-service-client/lib/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var axios_1 = require("axios");
var FormData = require("form-data");
// sample header= "multipart/form-data;boundary=Boundary_1"
// get the part after "boundary=" and before any subsequent ;
var extractMultipartBoundary = /.*;boundary=(Boundary.*);?.*/;
var extractMultipartFileName = /Content-Disposition: form-data; name="([^"]+)"/;
// eveything between { } including newlines. [^] is like . but matches newline
var extractJSONContent = /(\{[^]*\})/;
var extractCQLInclude = /include .* called (.*)/g;
var Client = /** @class */ (function () {
    function Client(serviceUrl) {
        this.url = serviceUrl;
        this.headers = {
            "Content-Type": "application/cql",
            Accept: "application/elm+json"
        };
    }
    /**
     * Function that requests web_service to convert the cql into elm.
     * @param cqlLibraries - object containing CqlLibraries that is the input to the function.
     * @return The resulting elm translation of the cql libraries.
     */
    Client.prototype.convertCQL = function (cqlLibraries) {
        return __awaiter(this, void 0, void 0, function () {
            var formdata, headers;
            var _this = this;
            return __generator(this, function (_a) {
                formdata = new FormData();
                Object.keys(cqlLibraries).forEach(function (key) {
                    var cqlLibrary = cqlLibraries[key];
                    if (cqlLibrary.cql) {
                        formdata.append("" + key, cqlLibrary.cql);
                    }
                });
                if (formdata.getHeaders)
                    headers = formdata.getHeaders();
                return [2 /*return*/, axios_1.default
                        .post(this.url, formdata, { headers: headers })
                        .then(function (elm) { return _this.handleElm(elm); })
                        .catch(function (error) {
                        var _a, _b;
                        if (((_a = error.response) === null || _a === void 0 ? void 0 : _a.status) === 400 && ((_b = error.response) === null || _b === void 0 ? void 0 : _b.data.library))
                            return _this.handleElm(error.response.data);
                        else
                            return error;
                    })];
            });
        });
    };
    Client.prototype.convertBasicCQL = function (cql) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // Connect to web service
                return [2 /*return*/, axios_1.default
                        .post(this.url, cql, {
                        headers: this.headers
                    })
                        .then(function (response) {
                        return response.data;
                    })
                        .catch(function (error) {
                        var _a, _b;
                        if (((_a = error.response) === null || _a === void 0 ? void 0 : _a.status) === 400 && ((_b = error.response) === null || _b === void 0 ? void 0 : _b.data.library))
                            return error.response.data;
                        else
                            return error;
                    })];
            });
        });
    };
    Client.prototype.handleElm = function (elm) {
        var header = elm.headers['content-type'];
        var boundary = '';
        if (header) {
            // sample header= "multipart/form-data;boundary=Boundary_1"
            var result = extractMultipartBoundary.exec(header);
            boundary = result ? "--" + result[1] : '';
        }
        var elmLibraries = {};
        return elm.data.split(boundary).reduce(function (oldArray, line) {
            var body = extractJSONContent.exec(line);
            if (body) {
                var elmName = extractMultipartFileName.exec(line);
                if (elmName)
                    oldArray[elmName[1]] = JSON.parse(body[1]);
            }
            return oldArray;
        }, elmLibraries);
    };
    return Client;
}());
exports.Client = Client;
