// cql-execution/lib/elm/expressions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.doContains = exports.doExcept = exports.doIncludes = exports.doIntersect = exports.doProperIncludes = exports.doAfter = exports.doUnion = exports.doBefore = void 0;
__exportStar(require("./expression"), exports);
__exportStar(require("./aggregate"), exports);
__exportStar(require("./arithmetic"), exports);
__exportStar(require("./clinical"), exports);
__exportStar(require("./comparison"), exports);
__exportStar(require("./conditional"), exports);
__exportStar(require("./datetime"), exports);
__exportStar(require("./declaration"), exports);
__exportStar(require("./external"), exports);
__exportStar(require("./instance"), exports);
__exportStar(require("./interval"), exports);
__exportStar(require("./list"), exports);
__exportStar(require("./literal"), exports);
__exportStar(require("./logical"), exports);
__exportStar(require("./message"), exports);
__exportStar(require("./nullological"), exports);
__exportStar(require("./parameters"), exports);
__exportStar(require("./quantity"), exports);
__exportStar(require("./query"), exports);
__exportStar(require("./ratio"), exports);
__exportStar(require("./reusable"), exports);
__exportStar(require("./string"), exports);
__exportStar(require("./structured"), exports);
__exportStar(require("./type"), exports);
__exportStar(require("./overloaded"), exports);
// Re-exporting interval functions as overrides to avoid ambiguity
// https://stackoverflow.com/questions/41293108/how-to-do-re-export-with-overrides
// TODO: we should improve this by perhaps renaming and reworking these functions
// it's a bit confusing right now giving the interval exports precedence over the others
const interval_1 = require("./interval");
Object.defineProperty(exports, "doBefore", { enumerable: true, get: function () { return interval_1.doBefore; } });
Object.defineProperty(exports, "doUnion", { enumerable: true, get: function () { return interval_1.doUnion; } });
Object.defineProperty(exports, "doAfter", { enumerable: true, get: function () { return interval_1.doAfter; } });
Object.defineProperty(exports, "doProperIncludes", { enumerable: true, get: function () { return interval_1.doProperIncludes; } });
Object.defineProperty(exports, "doIntersect", { enumerable: true, get: function () { return interval_1.doIntersect; } });
Object.defineProperty(exports, "doIncludes", { enumerable: true, get: function () { return interval_1.doIncludes; } });
Object.defineProperty(exports, "doExcept", { enumerable: true, get: function () { return interval_1.doExcept; } });
Object.defineProperty(exports, "doContains", { enumerable: true, get: function () { return interval_1.doContains; } });
